source/Fleet.cpp

```diff
@@ -503,9 +517,27 @@ const System *Fleet::Enter(const System &system, Ship &ship, const System *sourc
 	// Choose which system this ship is coming from.
 	if(!source)
 	{
+		vector<int> systemWeights;
+		int totalWeight = 0;
+		for(const System *link : system.Links())
+		{
+			bool access = unconstrained || ship.GetGovernment()->IsRestrictedFrom(*link);
+			systemWeights.emplace_back(access);
+			totalWeight += access;
+		}
 		auto it = system.Links().cbegin();
-		advance(it, Random::Int(system.Links().size()));
-		source = *it;
+		int choice = Random::Int(totalWeight);
+		for(unsigned i = 0; i < systemWeights.size(); ++i, ++it)
+		{
+			if(!unconstrained && !ship.GetGovernment()->IsRestrictedFrom(*(*it)))
+				continue;
```

Comment:
You already checked this and inserted 0 if it can't use that system. Don't check it again.
```suggestion
```